Introduce SpriteSequence, a covariant supertype of SpriteList.

sjrd opened the pull request:
This is done by analogy to `collections.abc.Sequence`, which is a covariant supertype of `list`.

Before this commit, many parts of the codebase used `SpriteList`s without type arguments (defaulting to `Unknown`). That was the only way to allow reasonable usages of the given methods and attributes. However, doing so results in weaker typing.

Using `SpriteSequence`, we can add correct type arguments to almost all of the references that were using `SpriteList`s before.

The only missing pieces are `Scene` and `TileMap`. Unfortunately, their APIs are fundamentally unsound wrt. the type arguments of their `SpriteList`s. We cannot make it sound without breaking their APIs, so we do not change them.

As a bonus, we can now create lists of `SpriteList`s with varying type arguments, and generically call `draw` or `update` on them. Previously, the only common supertype of `SpriteList[A]` and `SpriteList[B]` was `object`, which meant it was not possible to call those methods on them.

In a sense, that ability mostly subsumes the convenience provided by `Scene`. A `list[SpriteSequence[BasicSprite]]` is almost as convenient, while being type-safe.

---

This was prompted by seeing students struggle with the absence of a common supertype for `SpriteList[A]` and `SpriteList[B]`. Students resorted to `cast`s or `Any`s (or smuggled their way out of that using `Scene`) to work around the issue. With `SpriteSequence`, this shouldn't be an issue anymore.

It turns out that the introduction of `SpriteSequence` allows to more safely type many APIs of Arcade, so hopefully this addition carries its weight.

sjrd committed:
Remove redundant call to `clear()` in `remove_from_sprite_lists()`.

As we are just out of a while loop whose stopping condition is that
`sprite_lists` is empty, calling `clear()` on it is redundant.

sjrd committed:
Concentrate updates to `BasicSprite.sprite_lists` in a pair of methods.

There was already a method `register_sprite_list` to handle all
additions to `sprite_lists`. We add a corresponding method
`_unregister_sprite_list` to handle removals.

We make the typings of these methods stricter, so that we can
enforce the correct typing invariant on `sprite_lists`. We also
make that invariant clearer in a comment.

`sprite_lists` is unfortunately unsafely visible to everyone.
So a user of the class could still violate the invariants. At least
now the *intended* usage is safe.

sjrd committed:
Remove dead code attribute Sprite._sprite_list.

sjrd committed:
Fix the type signature of `get_closest_sprite`.

This is similar to the fix done to `check_for_collision_with_list`
done in c387717ffe20f05accf3f906a86b7bbf3fbdd12d.

sjrd committed:
A few better types in arcade.future and arcade.particles.

Adding type parameters to some `SpriteList`s. One allows to get rid
of a cast.

sjrd committed:
Introduce SpriteSequence, a covariant supertype of SpriteList.

This is done by analogy to `collections.abc.Sequence`, which is a
covariant supertype of `list`.

Before this commit, many parts of the codebase used `SpriteList`s
without type arguments (defaulting to `Unknown`). That was the only
way to allow reasonable usages of the given methods and attributes.
However, doing so results in weaker typing.

Using `SpriteSequence`, we can add correct type arguments to almost
all of the references that were using `SpriteList`s before.

The only missing pieces are `Scene` and `TileMap`. Unfortunately,
their APIs are fundamentally unsound wrt. the type arguments of
their `SpriteList`s. We cannot make it sound without breaking their
APIs, so we do not change them.

As a bonus, we can now create lists of `SpriteList`s with varying
type arguments, and generically call `draw` or `update` on them.
Previously, the only common supertype of `SpriteList[A]` and
`SpriteList[B]` was `object`, which meant it was not possible to
call those methods on them.

In a sense, that ability mostly subsumes the convenience provided
by `Scene`. A `list[SpriteSequence[BasicSprite]]` is almost as
convenient, while being type-safe.

pushfoo commented:
> The only missing pieces are `Scene` and `TileMap`. Unfortunately, their APIs are fundamentally unsound wrt. the type arguments of their SpriteLists.

@DragonMoffon and I want to change `Scene`. I don't make use of `TileMap` that much, so I don't have a strong opinion on it.

pushfoo commented on arcade/sprite_list/spatial_hash.py:
```diff
@@ -1,13 +1,54 @@
+    See SpatialHash for more details.
+    """
+
+    @abstractmethod
```
As a heads up, Python's generics are kind of a mess compared to those in the Java ecosystem and anything related to it (including your Scala projects :eyes:).

I'm confused as to whether/how this actually works. My reading is that `typing.Protocol` may be a better fit since:

1. [`abc.abstractmethod` requires classes to have `ABCMeta` as their metaclass](https://docs.python.org/3/library/abc.html#abc.abstractmethod)
2. `Generic` has no metaclass at all since it's [defined as a C-backed type in `_typingmodule.c`](https://github.com/python/cpython/blob/00cf5eacc52d4790c23a24c835fb6b048cb27d8c/Modules/_typingmodule.c#L62)
3. We may want to support different spatial structures (quad tree, etc) in the future

pushfoo replied:
Update: At least on Python 3.13, `typing.Generic` doesn't work with `abc.abstractmethod` at all :(

Here's a cut-down version of the output + [the gist I wrote](https://gist.github.com/pushfoo/8a01bc61de554b5b11de8e86ef28c5c4):
```
Trying to instantiate GenericBase...
  FAIL: Did NOT raise a TypeError!

Trying to instantiate GenericSub...
  FAIL: Did NOT raise a TypeError!
```
```python
import abc
from typing import Collection, TypeVar, Generic

T_co = TypeVar('T_co', covariant=True)


def does_it_type_error(t: Generic[T_co]) -> Generic[T_co]:
    """A decorator which tries to instantiate the passed class immediately."""

    name = t.__name__
    try:
        print(F"Trying to instantiate {name}...")
        t()
        print(f"  FAIL: Did NOT raise a TypeError!\n")
    except TypeError as _:
        print(f"  PASS: Raised a TypeError.\n")
    finally:
        return t


@does_it_type_error
class GenericBase(Generic[T_co]):

    @abc.abstractmethod
    def local_method(self) -> int:
        ...


@does_it_type_error
class GenericSub(GenericBase[T_co]):
    ...
```

sjrd replied:
Huh, gosh, indeed it needed `Protocol`. It wasn't checking nearly as much as I thought it was just with `Generic`. Adding `Protocol` flagged the use of `set[SpriteType_co]` in covariant position, which I replaced by `collections.abc.Set`.

Coming from Scala, as you noticed, I'm generally pleasantly surprised by what the Python type system can express these days. But I sure get confused with Python's idea of what an abstract class/interface/ABC is.

pushfoo replied:
> But I sure get confused with Python's idea of what an abstract class/interface/ABC is.

So does Python, sadly. That's not your fault.

pushfoo commented on arcade/sprite_list/spatial_hash.py:
```diff
@@ -1,13 +1,54 @@
+    This is useful when the SpriteType is in covariant position.
+
+    See SpatialHash for more details.
+    """
```
Maybe something like this? The wording's less a concern to me than getting the cross-refs solid.
```suggestion
    """A read-only view of a :py:class:`.SpatialHash` which helps preserve safety.

    This works like the read-only views of Python's built-in :py:class:`dict` and other types. As an every-day user, it means
    you can put subtypes of the annotated type into the `SpatialHash` or a `SpriteList`, but not superclasses.
    
    This ensures predicable behavior via type safety in cases where:

    #. A spatial hash is annotated with a specific type
    #. The object could be modified outside the original context to add a broader type
   
    Advanced users who want more information on the specifics should see the comments of :py:class:`~arcade.sprite_list.SpriteList`.
    """
```

sjrd replied:
I took your suggestion but I amended a few things. LMK if you disagree.

pushfoo replied:
All good. We can clean it up later. Ty!

pushfoo commented on arcade/sprite_list/sprite_list.py:
```diff
@@ -39,8 +40,109 @@
+    SpriteLists of different sprite types.
+
+    See SpriteList for more details.
+    """
```
```suggestion
    """A read-only view of a :py:class:`.SpriteList`.

    Like other generics such as :py:class:`collections.abc.Sequence`, a `SpriteSequence` requires
    sprites be of a covariant type relative to their annotated type.
    
    See :py:class:`.SpriteList` for more details.
    """
```

pushfoo commented on arcade/sprite_list/spatial_hash.py:
```diff
@@ -1,13 +1,54 @@
+        """
+        Return sprites in the same buckets as the given rectangle.
+
+        Args:
+            rect: The rectangle to check (left, right, bottom, top)
```
If we're adding this, we may as well correct it.
```suggestion
    .. tip:: Use :py:mod:`arcade.types.rect`'s helper functions to create rectangle objects!

    Args:
        rect: The rectangle to check as a :py:class:`~arcade.types.rect.Rect` object.
```

pushfoo commented on tests/unit/spritelist/test_spritesequence.py:
```diff
@@ -0,0 +1,26 @@
+    sprite_list2: arcade.SpriteList[_CustomSpriteSolidColor] = arcade.SpriteList()
+    sprite_list2.append(_CustomSpriteSolidColor(16, 16, color=(255, 0, 0, 1)))
+
+    # It really is a SpriteList with a good type; this would not typecheck otherwise
+    custom_sprite: _CustomSpriteSolidColor = sprite_list2[0] # assert_type
```
Are you running this type check manually? My understanding is that we don't run type checks against the test folder at all. See:

https://github.com/pythonarcade/arcade/blob/94c42806cf63df636e2ae6eb2e8a0ceb25e31225/pyproject.toml#L140-L150

sjrd replied:
I don't think I ran pyright, but I did run mypy manually.

pushfoo replied:
We auto-run it in our side for build CI. As an important note, mypy refuses to support for `NamedTuple` at all. Since pyglet uses `NamedTuple` in pyglet's `math.Vec*` and `Mat*` classes:
* they mark `pyglet.math` as an excluded module in their mypy config
* we use pyright

pushfoo reviewed (changes requested):
Thank you for starting taking a crack at this. I appreciate it since it's sorely needed (see #2580 and similar :sweat_smile:).

For the moment, I'm a little confused on some `abc`-related details as in [this comment](https://github.com/pythonarcade/arcade/pull/2647#discussion_r2041096869). If nobody beats me to giving this a closer look, I may need a few days before I can give it the full attention it deserves.

@eruvanos btw, do you think we should add a typing-specific tests folder or similar as an exception to the tests folder being ignored?

sjrd reviewed (commented):
Thanks for the review. I addressed the comments so far.

I'm not sure what your squashing policy is. For now I added a comment addressing the review, but I'm also happy to squash it with the parent commit.

sjrd commented:
@pushfoo May I suggest a ping here? Is there anything else you'd like me to do? Thank you.

eruvanos commented:
@pushfoo Regarding the typed test folder: if we have tests that benefit from such a setup I am fine, I wonder if it makes our test structure more complicated. 

Regarding Scene and TileMap: we can iterate on them after we merge this PR. I just created a small code base over easter which uses Scene and TileMap, so I could doublecheck the improved versions.

@sjrd we do not have a strict squashing policy, I prefer squashed PRs to reduce the amount of commits we have to read through during write up of release notes. 
Talking about release notes, could you add your change to the release notes of `3.2 (unreleased)`?

sjrd commented:
Thanks. I squashed the last two commits, and added the entry to the changelog. I had to rebase on top of #2652 not to have conflicts. The only changes besides rebasing are the `CHANGELOG` file.

eruvanos commented:
@pushfoo are you fine with the changes? I think we could move on with this and iterate on Scene and Tilemap (and typed tests afterwards). What is your opinion?

pushfoo commented on arcade/sprite_list/collision.py:
```diff
@@ -365,7 +365,7 @@ def get_sprites_in_rect(rect: Rect, sprite_list: SpriteList[SpriteType]) -> List
     """
     if __debug__:
-        if not isinstance(sprite_list, SpriteList):
+        if not isinstance(sprite_list, SpriteSequence):
```
This is using `Collection` rather than protocol so I think this may work without a perf hit / lying the way `Protocol` isinstance checks work.

pushfoo reviewed (approved):
Sorry for the delay. I'll say let's merge it to keep velocity and revisit if we need to. Ty for your contribution and your patience.
